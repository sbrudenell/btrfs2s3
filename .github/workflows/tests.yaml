name: tests

permissions:
  contents: read

on:
  pull_request:
  push:
    branches: [master]

jobs:
  tox:
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-24.04, ubuntu-22.04]
        py: ["3.9", "3.10", "3.11", "3.12", "3.13"]
    steps:
    - uses: actions/checkout@v4
    - uses: actions/setup-python@v5
      with:
        python-version: ${{ matrix.py }}
        cache: pip
    - run: |
        python -m pip install --upgrade tox
        sudo -E env PATH=$PATH tox

  pre-commit:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        py: ["3.9", "3.10", "3.11", "3.12", "3.13"]
    steps:
    - uses: actions/checkout@v4
    - uses: actions/setup-python@v5
      with:
        python-version: ${{ matrix.py }}
        cache: pip
    - uses: pre-commit/action@v3.0.1
