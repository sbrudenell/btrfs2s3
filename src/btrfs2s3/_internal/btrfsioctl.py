# btrfs2s3 - maintains a tree of differential backups in object storage.
#
# Copyright (C) 2025 Steven Brudenell and other contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

from contextlib import contextmanager
from ctypes import c_char
from ctypes import c_int64
from ctypes import c_uint8
from ctypes import c_uint32
from ctypes import c_uint64
from ctypes import POINTER
from ctypes import sizeof
from ctypes import Structure
from ctypes import Union as CtypesUnion
from enum import IntFlag
from fcntl import ioctl
import os
import os.path
from pathlib import Path
import re
from subprocess import check_output
from typing import cast
from typing import NamedTuple
from typing import overload
from typing import Protocol
from typing import runtime_checkable
from typing import TYPE_CHECKING
from typing import Union

if TYPE_CHECKING:
    from collections.abc import Iterator
    from ctypes import _CData
    from typing import AnyStr

    from typing_extensions import TypeAlias


_IOC_NRBITS = 8
_IOC_TYPEBITS = 8
_IOC_SIZEBITS = 14
_IOC_DIRBITS = 2

_IOC_NONE = 0
_IOC_WRITE = 1
_IOC_READ = 2


_machine = os.uname().machine

if _machine.startswith("parisc"):  # pragma: no cover
    _IOC_WRITE = 2
    _IOC_READ = 1
elif _machine == "alpha" or _machine.startswith(
    ("mips", "sparc", "ppc", "powerpc")
):  # pragma: no cover
    _IOC_SIZEBITS = 13
    _IOC_DIRBITS = 3
    _IOC_NONE = 1
    _IOC_READ = 2
    _IOC_WRITE = 4


_IOC_NRSHIFT = 0
_IOC_TYPESHIFT = _IOC_NRSHIFT + _IOC_NRBITS
_IOC_SIZESHIFT = _IOC_TYPESHIFT + _IOC_TYPEBITS
_IOC_DIRSHIFT = _IOC_SIZESHIFT + _IOC_SIZEBITS

IOCTL_MAGIC = 0x94


def _ioc(_dir: int, nr: int, size: _CData | type[_CData] | int) -> int:
    return (
        (_dir << _IOC_DIRSHIFT)
        | (IOCTL_MAGIC << _IOC_TYPESHIFT)
        | (nr << _IOC_NRSHIFT)
        | (size if isinstance(size, int) else sizeof(size) << _IOC_SIZESHIFT)
    )


def _ior(nr: int, size: _CData | type[_CData]) -> int:
    return _ioc(_IOC_READ, nr, size)


def _iow(nr: int, size: _CData | type[_CData]) -> int:
    return _ioc(_IOC_WRITE, nr, size)


VOL_NAME_MAX = 255
SUBVOL_NAME_MAX = 4039
UUID_SIZE = 16
FS_TREE_OBJECTID = 5
FIRST_FREE_OBJECTID = 256
NULL_UUID = b"\0" * UUID_SIZE


class SubvolArgsFlag(IntFlag):
    ReadOnly = 1 << 1
    QgroupInherit = 1 << 2
    DeviceSpecById = 1 << 3
    SubvolSpecById = 1 << 4


class SubvolFlag(IntFlag):
    ReadOnly = 1 << 0


class _VolArgsV2UnionId(CtypesUnion):
    _fields_ = (
        ("name", c_char * (SUBVOL_NAME_MAX + 1)),
        ("devid", c_uint64),
        ("subvolid", c_uint64),
    )


class VolArgsV2(Structure):
    _fields_ = (
        ("fd", c_int64),
        ("transid", c_uint64),
        ("flags", c_uint64),
        # qgroup args go here, but we don't use them
        ("_unused", c_uint64 * 4),
        ("_u", _VolArgsV2UnionId),
    )
    _anonymous_ = ("_u",)


assert sizeof(VolArgsV2) == 4096  # noqa: PLR2004


class TimeSpec(Structure):
    _fields_ = (("sec", c_uint64), ("nsec", c_uint32))

    def to_float(self) -> float:
        return cast("int", self.sec) + cast("int", self.nsec) * 1e-9


class SendFlag(IntFlag):
    NoFileData = 0x1
    OmitStreamHeader = 0x2
    OmitEndCmd = 0x4
    Version = 0x8
    Compressed = 0x10


class SendArgs(Structure):
    _fields_ = (
        ("send_fd", c_int64),
        ("_clone_sources_count", c_uint64),
        ("_clone_sources", POINTER(c_uint64)),
        ("parent_root", c_uint64),
        ("flags", c_uint64),
        ("version", c_uint32),
        ("_reserved", c_uint8 * 28),
    )


assert sizeof(SendArgs) == (72 if sizeof(POINTER(c_uint64)) == 8 else 68)  # noqa: PLR2004


class SubvolInfo(NamedTuple):
    id: int = 0
    name: str = ""
    parent_id: int = 0
    dir_id: int = 0
    generation: int = 0
    flags: int = 0
    uuid: bytes = NULL_UUID
    parent_uuid: bytes | None = None
    received_uuid: bytes | None = None
    ctransid: int = 0
    otransid: int = 0
    stransid: int = 0
    rtransid: int = 0
    ctime: float = 0.0
    otime: float = 0.0
    stime: float = 0.0
    rtime: float = 0.0


class SubvolInfoStruct(Structure):
    _fields_ = (
        ("id", c_uint64),
        ("name", c_char * (VOL_NAME_MAX + 1)),
        ("parent_id", c_uint64),
        ("dir_id", c_uint64),
        ("generation", c_uint64),
        ("flags", c_uint64),
        ("uuid", c_uint8 * UUID_SIZE),
        ("parent_uuid", c_uint8 * UUID_SIZE),
        ("received_uuid", c_uint8 * UUID_SIZE),
        ("ctransid", c_uint64),
        ("otransid", c_uint64),
        ("stransid", c_uint64),
        ("rtransid", c_uint64),
        ("ctime", TimeSpec),
        ("otime", TimeSpec),
        ("stime", TimeSpec),
        ("rtime", TimeSpec),
        ("_reserved", c_uint64 * 8),
    )

    def to_tuple(self) -> SubvolInfo:
        return SubvolInfo(
            id=self.id,
            name=os.fsdecode(self.name),
            parent_id=self.parent_id,
            dir_id=self.dir_id,
            generation=self.generation,
            flags=self.flags,
            uuid=bytes(self.uuid),
            parent_uuid=bytes(self.parent_uuid)
            if bytes(self.parent_uuid) != NULL_UUID
            else None,
            received_uuid=bytes(self.received_uuid)
            if bytes(self.received_uuid) != NULL_UUID
            else None,
            ctransid=self.ctransid,
            otransid=self.otransid,
            stransid=self.stransid,
            rtransid=self.rtransid,
            ctime=self.ctime.to_float(),
            otime=self.otime.to_float(),
            stime=self.stime.to_float(),
            rtime=self.rtime.to_float(),
        )


IOC_SNAP_CREATE_V2 = _iow(23, VolArgsV2)
IOC_SUBVOL_CREATE_V2 = _iow(24, VolArgsV2)
IOC_SEND = _iow(38, SendArgs)
IOC_GET_SUBVOL_INFO = _ior(60, SubvolInfoStruct)
IOC_SNAP_DESTROY_V2 = _iow(63, VolArgsV2)


_PathLike: TypeAlias = Union[str, bytes, os.PathLike[str], os.PathLike[bytes]]


@overload
def _split(p: AnyStr) -> tuple[AnyStr, AnyStr]: ...
@overload
def _split(p: os.PathLike[AnyStr]) -> tuple[AnyStr, AnyStr]: ...
def _split(p: AnyStr | os.PathLike[AnyStr]) -> tuple[AnyStr, AnyStr]:
    return os.path.split(os.path.realpath(p))


@contextmanager
def opendir(path: _PathLike, dir_fd: int | None = None) -> Iterator[int]:
    fd = os.open(path, os.O_RDONLY, dir_fd=dir_fd)
    try:
        yield fd
    finally:
        os.close(fd)


def subvol_info(fd: int | _PathLike, /, *, dir_fd: int | None = None) -> SubvolInfo:
    if not isinstance(fd, int):
        with opendir(fd, dir_fd=dir_fd) as fd_:
            return subvol_info(fd_)
    info = SubvolInfoStruct()
    ioctl(fd, IOC_GET_SUBVOL_INFO, info)
    return info.to_tuple()


def create_snap(
    *,
    src: int | _PathLike,
    dst: _PathLike,
    src_dir_fd: int | None = None,
    dst_dir_fd: int | None = None,
    read_only: bool = False,
) -> None:
    if not isinstance(src, int):
        with opendir(src, dir_fd=src_dir_fd) as src_fd:
            create_snap(src=src_fd, dst=dst, dst_dir_fd=dst_dir_fd, read_only=read_only)
            return
    if dst_dir_fd is None:
        (dst_dir, dst) = _split(dst)
        with opendir(dst_dir) as dst_dir_fd:  # noqa: PLR1704
            create_snap(src=src, dst=dst, dst_dir_fd=dst_dir_fd, read_only=read_only)
            return
    flags = 0
    if read_only:
        flags |= SubvolArgsFlag.ReadOnly
    ioctl(
        dst_dir_fd,
        IOC_SNAP_CREATE_V2,
        VolArgsV2(fd=src, name=os.fsencode(dst), flags=flags),
    )


def create_subvol(path: _PathLike, /, *, dir_fd: int | None = None) -> None:
    if dir_fd is None:
        dir_, path = _split(path)
        with opendir(dir_) as dir_fd:  # noqa: PLR1704
            create_subvol(path, dir_fd=dir_fd)
            return
    ioctl(dir_fd, IOC_SUBVOL_CREATE_V2, VolArgsV2(name=os.fsencode(path)))


def get_kernel_send_proto() -> int:
    # /sys/fs/btrfs may not exist at all before the module is loaded. make sure
    # we raise an error in this case
    base_path = Path("/sys/fs/btrfs")
    if not base_path.exists():
        raise FileNotFoundError(base_path)  # pragma: no cover
    # send_stream_version was introduced in kernel 5.9. before that, only
    # version 1 was supported.
    path = Path("/sys/fs/btrfs/features/send_stream_version")
    if path.exists():
        return int(path.read_text())  # pragma: no cover
    return 1  # pragma: no cover


def get_userspace_send_proto() -> int:
    full_output = check_output(["btrfs", "version"], text=True)
    m = re.match(r"^btrfs-progs v(.*)$", full_output, re.MULTILINE)
    assert m is not None, full_output
    version_str = m.group(1)
    version = tuple(int(part) for part in version_str.split("."))
    if version >= (5, 19):
        return 2  # pragma: no cover
    return 1  # pragma: no cover


@runtime_checkable
class _HasFileno(Protocol):
    def fileno(self) -> int: ...


def send(
    *,
    src: int | _PathLike,
    dst: int | _HasFileno,
    proto: int | None = None,
    parent_id: int = 0,
    flags: int = 0,
) -> None:
    if not isinstance(src, int):
        with opendir(src) as fd:
            send(src=fd, dst=dst, proto=proto, parent_id=parent_id, flags=flags)
            return
    if not isinstance(dst, int):
        dst = dst.fileno()
    if proto is None:
        proto = 0
    else:
        flags |= SendFlag.Version
    ioctl(
        src,
        IOC_SEND,
        SendArgs(send_fd=dst, flags=flags, parent_root=parent_id, version=proto),
    )


@overload
def destroy_snap(*, dir_fd: int, snapshot_id: int) -> None: ...
@overload
def destroy_snap(path: _PathLike, /, *, dir_fd: int | None = None) -> None: ...
@overload
def destroy_snap(path: _PathLike, /, *, snapshot_id: int) -> None: ...
def destroy_snap(
    path: _PathLike | None = None,
    /,
    *,
    dir_fd: int | None = None,
    snapshot_id: int | None = None,
) -> None:
    if dir_fd is None:
        assert path is not None
        if snapshot_id is None:
            dir_, path = _split(path)
            with opendir(dir_) as dir_fd:  # noqa: PLR1704
                destroy_snap(path, dir_fd=dir_fd)
        else:
            with opendir(path) as dir_fd:
                destroy_snap(dir_fd=dir_fd, snapshot_id=snapshot_id)
        return
    args = VolArgsV2()
    if snapshot_id is not None:
        assert path is None
        args.flags |= SubvolArgsFlag.SubvolSpecById
        args.subvolid = snapshot_id
    else:
        assert path is not None
        args.name = os.fsencode(path)
    ioctl(dir_fd, IOC_SNAP_DESTROY_V2, args)
