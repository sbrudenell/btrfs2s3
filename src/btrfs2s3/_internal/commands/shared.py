# btrfs2s3 - maintains a tree of differential backups in object storage.
#
# Copyright (C) 2025 Steven Brudenell and other contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

"""Code shared between commands."""

from typing import TypedDict


class Args(TypedDict, total=False):
    help: str
    description: str
    epilog: str
