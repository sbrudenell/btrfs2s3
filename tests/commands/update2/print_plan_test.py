# btrfs2s3 - maintains a tree of differential backups in object storage.
#
# Copyright (C) 2024 Steven Brudenell and other contributors.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from __future__ import annotations

from functools import partial
from pathlib import Path
from typing import TYPE_CHECKING

import arrow
import pytest

from btrfs2s3._internal.btrfsioctl import SubvolInfo
from btrfs2s3._internal.commands.printing import print_plan
from btrfs2s3._internal.piper import filter_pipe
from btrfs2s3._internal.planner import AssessedBackup
from btrfs2s3._internal.planner import AssessedSnapshot
from btrfs2s3._internal.planner import Assessment
from btrfs2s3._internal.planner import assessment_to_actions
from btrfs2s3._internal.planner import BackupObject
from btrfs2s3._internal.planner import ObjectStat
from btrfs2s3._internal.planner import Remote
from btrfs2s3._internal.planner import SnapshotDir
from btrfs2s3._internal.planner import Source
from btrfs2s3._internal.resolver import Flags
from btrfs2s3._internal.resolver import KeepMeta
from btrfs2s3._internal.resolver import Reasons
from btrfs2s3._internal.util import backup_of_snapshot
from tests.conftest import AWS_COSTS

if TYPE_CHECKING:
    from rich.console import Console
    from types_boto3_s3.client import S3Client


def _t(t: str) -> float:
    return arrow.get(t).timestamp()


noop_pipe = partial(filter_pipe, [])


@pytest.fixture
def golden_asmt(s3: S3Client) -> Assessment:
    source1 = Source(
        path=Path("/path/to/source1"),
        info=SubvolInfo(
            id=12345, uuid=b"\01" * 16, ctransid=123, ctime=_t("2006-01-15")
        ),
        fd=-1,
    )
    source2 = Source(
        path=Path("/path/to/source2"),
        info=SubvolInfo(
            id=23456, uuid=b"\02" * 16, ctransid=234, ctime=_t("2006-02-15")
        ),
        fd=-1,
    )
    source3 = Source(
        path=Path("/path/to/source3"),
        info=SubvolInfo(
            id=34567, uuid=b"\x03" * 16, ctransid=345, ctime=_t("2006-03-15")
        ),
        fd=-1,
    )
    snap1_1 = SubvolInfo(
        id=source1.info.id + 1,
        uuid=b"\x11" * 16,
        parent_uuid=source1.info.uuid,
        ctransid=source1.info.ctransid - 3,
        ctime=source1.info.ctime - 86400,
    )
    snap1_2 = SubvolInfo(
        id=source1.info.id + 2,
        uuid=b"\x12" * 16,
        parent_uuid=source1.info.uuid,
        ctransid=source1.info.ctransid - 2,
        ctime=source1.info.ctime - 3600,
    )
    snap2_1 = SubvolInfo(
        id=source2.info.id + 3,
        uuid=b"\x21" * 16,
        parent_uuid=source2.info.uuid,
        ctransid=source2.info.ctransid - 3,
        ctime=source2.info.ctime - 86400,
    )
    snap2_2 = SubvolInfo(
        id=source2.info.id + 4,
        uuid=b"\x22" * 16,
        parent_uuid=source2.info.uuid,
        ctransid=source2.info.ctransid - 2,
        ctime=source2.info.ctime - 3600,
    )
    snap3_1 = SubvolInfo(
        id=source3.info.id + 5,
        uuid=b"\x31" * 16,
        parent_uuid=source3.info.uuid,
        ctransid=source3.info.ctransid - 3,
        ctime=source3.info.ctime - 86400,
    )
    snap3_2 = SubvolInfo(
        id=source3.info.id + 6,
        uuid=b"\x32" * 16,
        parent_uuid=source3.info.uuid,
        ctransid=source3.info.ctransid - 2,
        ctime=source3.info.ctime - 3600,
    )

    snapshot_dir1 = SnapshotDir(
        path=Path("/path/to/snapshot_dir1"),
        dir_fd=-1,
        snapshots=[("snap1_1", snap1_1), ("snap1_2", snap1_2)],
    )
    snapshot_dir2 = SnapshotDir(
        path=Path("/path/to/snapshot_dir2"),
        dir_fd=-1,
        snapshots=[
            ("snap2_1", snap2_1),
            ("snap2_2", snap2_2),
            ("snap3_1", snap3_1),
            ("snap3_2", snap3_2),
        ],
    )

    backup1_1 = backup_of_snapshot(snap1_1)
    backup1_2 = backup_of_snapshot(snap1_2, send_parent=backup1_1)
    backup2_1 = backup_of_snapshot(snap2_1)
    backup2_2 = backup_of_snapshot(snap2_2, send_parent=backup2_1)
    backup3_1 = backup_of_snapshot(snap3_1)
    backup3_2 = backup_of_snapshot(snap3_2, send_parent=backup3_1)

    obj1_1 = BackupObject(
        key=source1.get_backup_key(backup1_1),
        info=backup1_1,
        stat=ObjectStat(size=1 * 2**30, storage_class="DEEP_ARCHIVE"),
    )
    obj1_2 = BackupObject(
        key=source1.get_backup_key(backup1_2),
        info=backup1_2,
        stat=ObjectStat(size=2 * 2**30, storage_class="STANDARD"),
    )
    obj2_1 = BackupObject(
        key=source2.get_backup_key(backup2_1),
        info=backup2_1,
        stat=ObjectStat(size=3 * 2**30, storage_class="DEEP_ARCHIVE"),
    )
    obj2_2 = BackupObject(
        key=source2.get_backup_key(backup2_2),
        info=backup2_2,
        stat=ObjectStat(size=4 * 2**30),
    )
    obj3_1 = BackupObject(
        key=source3.get_backup_key(backup3_1),
        info=backup3_1,
        stat=ObjectStat(storage_class="DEEP_ARCHIVE"),
    )
    obj3_2 = BackupObject(
        key=source3.get_backup_key(backup3_2), info=backup3_2, stat=ObjectStat()
    )

    remote1 = Remote(
        name="remote1",
        s3=s3,
        bucket="bucket1",
        objects=[obj1_1, obj1_2, obj2_1, obj2_2],
        costs=AWS_COSTS,
    )
    remote2 = Remote(name="remote2", s3=s3, bucket="bucket2", objects=[obj3_1, obj3_2])

    meta = KeepMeta(reasons=Reasons.Preserved, time_spans={(_t("2006"), _t("2007"))})

    return Assessment(
        snapshots={
            snap1_1.uuid: AssessedSnapshot(
                source=source1, info=snap1_1, snapshot_dir=snapshot_dir1, meta=meta
            ),
            snap1_2.uuid: AssessedSnapshot(
                source=source1, info=snap1_2, snapshot_dir=snapshot_dir1, meta=meta
            ),
            snap2_1.uuid: AssessedSnapshot(
                source=source2, info=snap2_1, snapshot_dir=snapshot_dir2, meta=meta
            ),
            snap2_2.uuid: AssessedSnapshot(
                source=source2, info=snap2_2, snapshot_dir=snapshot_dir2, meta=meta
            ),
            snap3_1.uuid: AssessedSnapshot(
                source=source3, info=snap3_1, snapshot_dir=snapshot_dir2, meta=meta
            ),
            snap3_2.uuid: AssessedSnapshot(
                source=source3, info=snap3_2, snapshot_dir=snapshot_dir2, meta=meta
            ),
        },
        backups={
            (remote1, backup1_1.uuid): AssessedBackup(
                source=source1,
                remote=remote1,
                info=backup1_1,
                stat=obj1_1.stat,
                key=obj1_1.key,
                meta=meta,
                create_pipe=noop_pipe,
            ),
            (remote1, backup1_2.uuid): AssessedBackup(
                source=source1,
                remote=remote1,
                info=backup1_2,
                stat=obj1_2.stat,
                key=obj1_2.key,
                meta=meta,
                create_pipe=noop_pipe,
            ),
            (remote1, backup2_1.uuid): AssessedBackup(
                source=source2,
                remote=remote1,
                info=backup2_1,
                stat=obj2_1.stat,
                key=obj2_1.key,
                meta=meta,
                create_pipe=noop_pipe,
            ),
            (remote1, backup2_2.uuid): AssessedBackup(
                source=source2,
                remote=remote1,
                info=backup2_2,
                stat=obj2_2.stat,
                key=obj2_2.key,
                meta=meta,
                create_pipe=noop_pipe,
            ),
            (remote2, backup3_1.uuid): AssessedBackup(
                source=source3,
                remote=remote2,
                info=backup3_1,
                stat=obj3_1.stat,
                key=obj3_1.key,
                meta=meta,
                create_pipe=noop_pipe,
            ),
            (remote2, backup3_2.uuid): AssessedBackup(
                source=source3,
                remote=remote2,
                info=backup3_2,
                stat=obj3_2.stat,
                key=obj3_2.key,
                meta=meta,
                create_pipe=noop_pipe,
            ),
        },
    )


def test_print_plan(golden_asmt: Assessment, goldifyconsole: Console) -> None:
    print_plan(
        console=goldifyconsole,
        assessment=golden_asmt,
        actions=assessment_to_actions(golden_asmt),
    )


@pytest.mark.parametrize(
    "meta",
    [
        KeepMeta(),
        KeepMeta(reasons=Reasons.MostRecent),
        KeepMeta(reasons=Reasons.MostRecent, flags=Flags.New),
    ],
    ids=["delete", "mostrecent", "mostrecentnew"],
)
def test_alternate_keep_meta(
    golden_asmt: Assessment, goldifyconsole: Console, meta: KeepMeta
) -> None:
    # Pick an arbitrary, stable snapshot
    uuid, snap = sorted(golden_asmt.snapshots.items())[0]
    # Mark the snapshot
    golden_asmt.snapshots[uuid] = snap._replace(meta=meta)
    # And all associated backups
    for (remote, backup_uuid), backup in list(golden_asmt.backups.items()):
        if backup_uuid == uuid:
            golden_asmt.backups[(remote, uuid)] = backup._replace(meta=meta)

    print_plan(
        console=goldifyconsole,
        assessment=golden_asmt,
        actions=assessment_to_actions(golden_asmt),
    )


def test_send_ancestor(golden_asmt: Assessment, goldifyconsole: Console) -> None:
    # Pick an arbitrary, stable backup which has a send-parent
    (remote, child_uuid), child_backup = sorted(
        golden_asmt.backups.items(),
        key=lambda i: (i[1].info.send_parent_uuid is None, i[0][1], i[0][0].name),
    )[0]
    # Get its parent snapshot and backup
    parent_uuid = child_backup.info.send_parent_uuid
    assert parent_uuid is not None
    parent_snap = golden_asmt.snapshots[parent_uuid]
    parent_backup = golden_asmt.backups[(remote, parent_uuid)]
    # Snapshot is not kept
    golden_asmt.snapshots[parent_uuid] = parent_snap._replace(meta=KeepMeta())
    # Backup must be kept because it's the send-parent of another backup
    golden_asmt.backups[(remote, parent_uuid)] = parent_backup._replace(
        meta=KeepMeta(reasons=Reasons.SendAncestor, other_uuids={child_uuid})
    )

    print_plan(
        console=goldifyconsole,
        assessment=golden_asmt,
        actions=assessment_to_actions(golden_asmt),
    )
