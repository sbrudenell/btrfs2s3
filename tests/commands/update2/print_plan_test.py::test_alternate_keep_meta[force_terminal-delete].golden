[3msnapshots                                                                            [0m
 ─────────────────────────────────────────────────────────────────────────────────── 
 [1m [0m[1mpath                            [0m[1m [0m [1m [0m[1mctime              [0m[1m [0m [1m [0m[1mctransid[0m[1m [0m [1m [0m[1mpreserve     [0m[1m [0m 
 ─────────────────────────────────────────────────────────────────────────────────── 
  [31m💀/path/to/snapshot_dir1/snap1_1[0m   [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m14[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m  [36m [0m[36m120     [0m[36m [0m  [1;91m<not keeping>[0m  
 [2m [0m[2m/path/to/snapshot_dir1/snap1_2  [0m[2m [0m [2m [0m[2;34m2006[0m[2;34m-[0m[2;34m01[0m[2;34m-[0m[2;34m14[0m[2mT[0m[2;35m23[0m[2;35m:[0m[2;35m00[0m[2;35m:[0m[2;35m00[0m[2m [0m [2;36m [0m[2;36m121     [0m[2;36m [0m [2m [0m[2;34m2006[0m[2m yearly  [0m[2m [0m 
  /path/to/snapshot_dir2/snap2_1     [34m2006[0m[34m-[0m[34m02[0m[34m-[0m[34m14[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m  [36m [0m[36m231     [0m[36m [0m  [34m2006[0m yearly    
 [2m [0m[2m/path/to/snapshot_dir2/snap2_2  [0m[2m [0m [2m [0m[2;34m2006[0m[2;34m-[0m[2;34m02[0m[2;34m-[0m[2;34m14[0m[2mT[0m[2;35m23[0m[2;35m:[0m[2;35m00[0m[2;35m:[0m[2;35m00[0m[2m [0m [2;36m [0m[2;36m232     [0m[2;36m [0m [2m [0m[2;34m2006[0m[2m yearly  [0m[2m [0m 
  /path/to/snapshot_dir2/snap3_1     [34m2006[0m[34m-[0m[34m03[0m[34m-[0m[34m14[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m  [36m [0m[36m342     [0m[36m [0m  [34m2006[0m yearly    
 [2m [0m[2m/path/to/snapshot_dir2/snap3_2  [0m[2m [0m [2m [0m[2;34m2006[0m[2;34m-[0m[2;34m03[0m[2;34m-[0m[2;34m14[0m[2mT[0m[2;35m23[0m[2;35m:[0m[2;35m00[0m[2;35m:[0m[2;35m00[0m[2m [0m [2;36m [0m[2;36m343     [0m[2;36m [0m [2m [0m[2;34m2006[0m[2m yearly  [0m[2m [0m 
 ─────────────────────────────────────────────────────────────────────────────────── 

backups
├── ☁️remote1
│   ├── 💀[31msource1.ctim2006-01-14T00:00:00+00:00.ctid120.uuid11111111-1111-1111-1111-111…[0m
│   │   [1;91m<not keeping>[0m [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m14[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m [36m120[0m [33m1.1 GB[0m [33mDEEP_ARCHIVE[0m [33m0.01/P1M[0m
│   │   └── [32msource1.ctim2006-01-14T23:00:00+00:00.ctid121.uuid12121212-1212-1212-1212-1…[0m
│   │       [34m2006[0m yearly [34m2006[0m[34m-[0m[34m01[0m[34m-[0m[34m14[0mT[35m23[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m [36m121[0m [33m2.1 GB[0m [33mSTANDARD[0m [33m0.05/P1M[0m
│   └── [32msource2.ctim2006-02-14T00:00:00+00:00.ctid231.uuid21212121-2121-2121-2121-21212…[0m
│       [34m2006[0m yearly [34m2006[0m[34m-[0m[34m02[0m[34m-[0m[34m14[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m [36m231[0m [33m3.2 GB[0m [33mDEEP_ARCHIVE[0m [33m0.02/P1M[0m
│       └── [32msource2.ctim2006-02-14T23:00:00+00:00.ctid232.uuid22222222-2222-2222-2222-2…[0m
│           [34m2006[0m yearly [34m2006[0m[34m-[0m[34m02[0m[34m-[0m[34m14[0mT[35m23[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m [36m232[0m [33m4.3 GB[0m
└── ☁️remote2
    └── [32msource3.ctim2006-03-14T00:00:00+00:00.ctid342.uuid31313131-3131-3131-3131-31313…[0m
        [34m2006[0m yearly [34m2006[0m[34m-[0m[34m03[0m[34m-[0m[34m14[0mT[35m00[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m [36m342[0m [33mDEEP_ARCHIVE[0m
        └── [32msource3.ctim2006-03-14T23:00:00+00:00.ctid343.uuid32323232-3232-3232-3232-3…[0m
            [34m2006[0m yearly [34m2006[0m[34m-[0m[34m03[0m[34m-[0m[34m14[0mT[35m23[0m[35m:[0m[35m00[0m[35m:[0m[35m00[0m [36m343[0m

actions to take:
[32m [0m[32m📝️ rename:[0m[32m [0m[32m [0m[32m/path/to/snapshot_dir1/snap1_2[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32msource1.2006-01-14T23:00:00+00:00.121[0m[32m [0m
[2;32m [0m[2;32m📝️ rename:[0m[2;32m [0m[2;32m [0m[2;32m/path/to/snapshot_dir2/snap2_1[0m[2;32m [0m[2;32m [0m[2;32m->[0m[2;32m [0m[2;32m [0m[2;32msource2.2006-02-14T00:00:00+00:00.231[0m[2;32m [0m
[32m [0m[32m📝️ rename:[0m[32m [0m[32m [0m[32m/path/to/snapshot_dir2/snap2_2[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32msource2.2006-02-14T23:00:00+00:00.232[0m[32m [0m
[2;32m [0m[2;32m📝️ rename:[0m[2;32m [0m[2;32m [0m[2;32m/path/to/snapshot_dir2/snap3_1[0m[2;32m [0m[2;32m [0m[2;32m->[0m[2;32m [0m[2;32m [0m[2;32msource3.2006-03-14T00:00:00+00:00.342[0m[2;32m [0m
[32m [0m[32m📝️ rename:[0m[32m [0m[32m [0m[32m/path/to/snapshot_dir2/snap3_2[0m[32m [0m[32m [0m[32m->[0m[32m [0m[32m [0m[32msource3.2006-03-14T23:00:00+00:00.343[0m[32m [0m
[31m [0m[31m💀 destroy:[0m[31m [0m[31m [0m[31m/path/to/snapshot_dir1/snap1_1[0m[31m [0m
[31m [0m[31m💀 delete:[0m[31m [0m[31m [0m[31m☁️remote1[0m[31m [0m[31m [0m[31msource1.ctim2006-01-14T00:00:00+00:00.ctid120.uuid11111111-1111…[0m[31m [0m

